lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -0,0 +1,568 @@
+  }
+
+  FailureOr<DeletionKind>
+  visitExternalOp(Operation *op,
+                  function_ref<void(Operation *)> addToWorklist) {
+    // TODO: we only have this to be able to write tests for this pass without
+    // having to add support for more operations for now, so it should be
+    // removed once it is not necessary anymore for writing tests
+    if (op->use_empty()) {
+      for (auto &operand : op->getOpOperands()) {
+        Value val = materializer.materialize(state.at(operand.get()),
+                                             op->getBlock(), op->getLoc());
+        if (!val)
+          return op->emitError("failed to materialize value");
+        operand.set(val);
+      }
+      return DeletionKind::Keep;
+    }
+
+    return visitUnhandledOp(op, addToWorklist);
+  }
+
+  FailureOr<DeletionKind>
+  visitOp(SetCreateOp op, function_ref<void(Operation *)> addToWorklist) {
+    SetVector<ElaboratorValue *> set;
+    for (auto val : op.getElements())
+      set.insert(state.at(val));
+
+    internalizeResult<SetValue>(op.getSet(), std::move(set),
+                                op.getSet().getType());
+    return DeletionKind::Delete;
+  }
+
+  FailureOr<DeletionKind>
+  visitOp(SetSelectRandomOp op, function_ref<void(Operation *)> addToWorklist) {
+    auto *set = cast<SetValue>(state.at(op.getSet()));
+
+    size_t selected;
+    std::uniform_int_distribution<size_t> dist(0, set->getSet().size() - 1);
+    if (auto intAttr =
+            op->getAttrOfType<IntegerAttr>("rtg.elaboration_custom_seed")) {
+      std::mt19937 customRng(intAttr.getInt());
+      selected = dist(customRng);
+    } else {
+      selected = dist(rng);
+    }
+
+    state[op.getResult()] = set->getSet()[selected];
+    return DeletionKind::Delete;
+  }
+
+  FailureOr<DeletionKind>
+  visitOp(SetDifferenceOp op, function_ref<void(Operation *)> addToWorklist) {
+    auto original = cast<SetValue>(state.at(op.getOriginal()))->getSet();
+    auto diff = cast<SetValue>(state.at(op.getDiff()))->getSet();
+
+    SetVector<ElaboratorValue *> result(original);
+    result.set_subtract(diff);
+
+    internalizeResult<SetValue>(op.getResult(), std::move(result),
+                                op.getResult().getType());
+    return DeletionKind::Delete;
+  }
+
+  FailureOr<DeletionKind>
+  dispatchOpVisitor(Operation *op,
+                    function_ref<void(Operation *)> addToWorklist) {
+    if (op->hasTrait<OpTrait::ConstantLike>()) {
+      SmallVector<OpFoldResult, 1> result;
+      auto foldResult = op->fold(result);
+      (void)foldResult; // Make sure there is a user when assertions are off.
+      assert(succeeded(foldResult) &&
+             "constant folder of a constant-like must always succeed");
+      auto attr = dyn_cast<TypedAttr>(result[0].dyn_cast<Attribute>());
+      if (!attr)
+        return op->emitError(
+            "only typed attributes supported for constant-like operations");
```

Comment:
Should probably add a constant op with a non-typed attribute to `rtgtest` to test this.